nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/schema/inference/SchemaInferenceUtil.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.schema.inference;
+
+import org.apache.nifi.components.AllowableValue;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.PropertyDescriptor.Builder;
+import org.apache.nifi.context.PropertyContext;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.schema.access.SchemaAccessStrategy;
+import org.apache.nifi.serialization.RecordSchemaCacheService;
+
+import java.util.function.Supplier;
+
+public class SchemaInferenceUtil {
+    public static final AllowableValue INFER_SCHEMA = new AllowableValue("infer-schema", "Infer Schema",
```

Comment:
Maybe add (and call) an overridden version of `AvroTypeUtil.extractAvroSchema()` that won't validate field names, it could be used in MergeRecord when `getSchemaText()` returns null, and in Record Writers when writing the schema to an attribute, e.g.

Reply:
Fair enough. Unfortunately, though, Avro does not allow you to skip field validation when creating a Field object - only when calling Parser#parse. This is unfortunately. But what we can do as a workaround is to determine that a field name is invalid, and then substitute in a valid field name and add the original field name as an alias. So "reporting.task.type" would become "reporting_task_type" and would have an alias of "reporting.task.type" since aliases' names are not validated. This would allow our parsing to work properly still. Will do so.